_posts/frameworks/2016-12-19-spring-restdocs.md

```diff
@@ -0,0 +1,274 @@
+---
+layout: article
+title: Documenting your REST API with Spring Rest Docs
+categories: [frameworks]
+modified: 2016-12-19
+author: tom
+tags: [spring, rest, api, documentation, mvc, controller, integration, test, docs, junit]
+comments: true
+ads: false
+image:
+  feature: 
+  teaser: teaser/books.jpg
+  thumb:
+---
+
+The first impression counts. When you're developing an API of any kind, chances are 
+that the first impression is gained from a look into the API docs. If that first impression
+fails to convince, developers will go on looking for another API they can use instead.
+ 
+## Why not Swagger?
+Looking for a tool to document a RESTful API, the first tool you probably come across
+is [Swagger](http://swagger.io/). Among other things, Swagger provides tooling for a lot of different programming
+languages and frameworks and allows automated creation of an API documentation and even of a 
+[web frontend that can interact with your API](http://swagger.io/swagger-ui/). Also, Swagger is well established as a tool
+supporting the development of RESTful APIs.
+
+But at least if you're familiar to Java, there's a compelling reason to use 
+[Spring Rest Docs](https://projects.spring.io/spring-restdocs/) instead of or at least 
+additionally to Swagger: Spring Rest Docs integrates directly into your integration tests.
+Tests will fail if you forget to document a field that you have just added to your API
+or if you removed a field that is still part your API docs. This way, your documentation 
+is always up-to-date with your implementation.
+
+This article explains the basics of Spring Rest Docs along
+the lines of some code examples. If you want to see it in action, you may want to check out 
+the [coderadar](https://github.com/reflectoring/coderadar) project on github.
+
+## Snippet-Generating Integration Tests
+The following code snippet shows a simple integration test of a Spring MVC controller that
+exposes a REST API to create a `project` resource.
+
+```java
+@Test
+public void createProjectSuccessfully() throws Exception {
+    ProjectResource projectResource = ...
+    mvc().perform(post("/projects")
+            .content(toJson(projectResource))
+            .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk())
+            .andDo(document("projects/create");
+}
+```
+
+Let's have a look at the details: `mvc()` is a helper method that creates a `MockMvc` 
+object that we use to submit a POST request to the URL `/projects`. The result of the
+request is passed into the `document()` method to automatically create documentation
+for the request. The `document()` method is statically imported from the class
+`MockMvcRestDocumentation` to keep the code readable.
```

Comment:
I like that you explain where the method is statically imported from. Most examples skip this part, which makes it hard to follow.